### DISCUSSION TITLE
Finding generalized form of equation from multiple traces?

### DISCUSSION THREAD

**Discussion by AlbanMor** [MilesCranmer/PySR#530] [Q&A]
Hi! 
First off, congratulation on this awesome package!

I was wondering if there is a way to use PySR to identify a generalized form of the equation ruling different traces. A simple example would be identifying the equation "a*x+b" from the following traces : y1 = 2*x1+3, y2= 3*x2+5, y3= -1*x3 + 4 etc.  

Thank you in advance!

**Comment by MilesCranmer** [comment]
Hi @AlbanMor,
Thanks for the kind words.

Very interesting idea. Right now there is not a built-in way to do this. But ideas appreciated.

Right now you could hack it – I think the easiest way to do something like this is to simply turn it into a single equation search (rather than 3), and pass the index of the equation (y1, y2, y3) as an additional feature. Then it could find something like this:

```
y = (2 > i) * 2 * x1 + (3 > i) * (i > 1) * 3 * x2 - (i > 2) * x3 + 2 + i
```

which as you can see gives you all `y1` (i=1), `y2` (i=2), and `y3` (i=3) in a single expression. Furthermore it will try to find the minimal equation that fits all three at once.

(You can pass `greater` as one of the operators)

I'm sure there's a smarter way to do this though... Ideas appreciated!

Cheers,
Miles

**Reply by AlbanMor** [reply]
Thank you so much for the quick reply. 
I'll give that a try but could you explain to me how that would diverge from solving separately for the three traces?

**Comment by AlbanMor** [comment]
I don't know the intricacies of your code but what I understand is that you first "evolve the operator tree" (sorry if i don't use the right terms) and then fit the constants values to the data. if these are separate steps, the fitting procedure could be done separately on each traces and return an average goodness of fit for that "operator tree". You could then judge how successful the "new evolved tree" is and continue to the next "evolution/mutation".  That would ensure that the same "operator tree" is used for all traces.

**Comment by Alicecomma** [comment]
One of the only hits to show up for this problem ('generalized form', 'fitting multiple datasets with a single expression'); it seems the documentation in PySR and SR.jl is split between calling these Template expressions (that you make highly restricted), Parametrized template expressions, Parametrized expressions, Class-dependent or Per-class problems.

This post is Jan 22, 2024 - a similar issue was solved here (Jun 24, 2024 -> Aug 25, 2025): https://github.com/MilesCranmer/SymbolicRegression.jl/issues/316 with "Template expressions" that are implemented in PySR with examples here: https://github.com/MilesCranmer/PySR/blob/225e0a568ff614238de04620fd9b91c2bdf34cc2/docs/src/examples.md#template-expressions and in Julia here: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/082c2dfe1b8a950aa12e9be3a003ecd7f629eb45/docs/src/examples.md with (dead) link to an example, probably similar to this: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/082c2dfe1b8a950aa12e9be3a003ecd7f629eb45/examples/parameterized_function.jl#L52

**Reply by MilesCranmer** [reply]
A "TemplateExpression" is the actual object name, while these others are more or less scientific terms for the type of expression you are looking for. i.e., a "parametrized expression" is not an internal object name but simply a more scientific _description_ of what we are doing. That one is also a TemplateExpression though. It's just flexible enough that you can build templates and also parametrized expressions with the same object.
